File: backend/app/db/models/expense.py
from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from sqlalchemy import Boolean, DateTime, Enum, ForeignKey, Integer, Numeric, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.db.models.enums import (
    AttachmentType,
    AuditResult,
    ExpenseApprovalAction,
    ExpenseStatus,
)
from app.db.models.mixins import TimestampMixin


class ExpenseClaim(Base, TimestampMixin):
    __tablename__ = "expense_claims"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    status: Mapped[ExpenseStatus] = mapped_column(
        Enum(ExpenseStatus, name="expense_status"), nullable=False
    )
    claim_no: Mapped[str | None] = mapped_column(String(64), unique=True, nullable=True)
    expense_type: Mapped[str] = mapped_column(
        String(64), default="travel", nullable=False
    )
    title: Mapped[str | None] = mapped_column(String(255), nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    total_amount: Mapped[Decimal | None] = mapped_column(
        Numeric(12, 2), nullable=True
    )
    currency: Mapped[str] = mapped_column(String(8), default="CNY", nullable=False)
    submitted_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    approved_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    audit_summary: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    reviewer_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    reviewed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    review_comment: Mapped[str | None] = mapped_column(Text, nullable=True)

    attachments: Mapped[list["ExpenseAttachment"]] = relationship(
        back_populates="claim", cascade="all, delete-orphan"
    )
    audit_items: Mapped[list["ExpenseAuditItem"]] = relationship(
        back_populates="claim", cascade="all, delete-orphan"
    )


class ExpenseAttachment(Base, TimestampMixin):
    __tablename__ = "expense_attachments"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    claim_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("expense_claims.id"), nullable=False
    )
    file_uri: Mapped[str] = mapped_column(String(512), nullable=False)
    file_name: Mapped[str] = mapped_column(String(255), nullable=False)
    file_type: Mapped[str] = mapped_column(String(64), nullable=False)
    size: Mapped[int | None] = mapped_column(Integer, nullable=True)
    attachment_type: Mapped[AttachmentType] = mapped_column(
        Enum(AttachmentType, name="attachment_type"), nullable=False
    )
    ocr_result: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    extracted_fields: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    ocr_confidence: Mapped[Decimal | None] = mapped_column(
        Numeric(5, 4), nullable=True
    )
    classification_source: Mapped[str | None] = mapped_column(String(32), nullable=True)

    claim: Mapped["ExpenseClaim"] = relationship(back_populates="attachments")


class ExpenseAuditItem(Base, TimestampMixin):
    __tablename__ = "expense_audit_items"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    claim_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("expense_claims.id"), nullable=False
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    result: Mapped[AuditResult] = mapped_column(
        Enum(
            AuditResult,
            name="audit_result",
            values_callable=lambda enum_cls: [item.value for item in enum_cls],
        ),
        nullable=False,
    )
    evidence: Mapped[str | None] = mapped_column(Text, nullable=True)

    claim: Mapped["ExpenseClaim"] = relationship(back_populates="audit_items")


class ExpenseApprovalLog(Base, TimestampMixin):
    __tablename__ = "expense_approval_logs"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    claim_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("expense_claims.id"), nullable=False
    )
    actor_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    action: Mapped[ExpenseApprovalAction] = mapped_column(
        Enum(ExpenseApprovalAction, name="expense_approval_action"), nullable=False
    )
    from_status: Mapped[str | None] = mapped_column(String(64), nullable=True)
    to_status: Mapped[str | None] = mapped_column(String(64), nullable=True)
    comment: Mapped[str | None] = mapped_column(Text, nullable=True)
    snapshot: Mapped[dict | None] = mapped_column(JSONB, nullable=True)


class TravelExpenseStandard(Base, TimestampMixin):
    __tablename__ = "travel_expense_standards"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    org_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("organizations.id"), nullable=True
    )
    city_tier: Mapped[str | None] = mapped_column(String(32), nullable=True)
    daily_limit: Mapped[Decimal | None] = mapped_column(Numeric(12, 2), nullable=True)
    single_trip_limit: Mapped[Decimal | None] = mapped_column(
        Numeric(12, 2), nullable=True
    )
    currency: Mapped[str] = mapped_column(String(8), default="CNY", nullable=False)
    effective_from: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    effective_to: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    metadata_: Mapped[dict | None] = mapped_column("metadata", JSONB, nullable=True)

File: frontend/src/types/expense.ts
export type ExpenseStatus =
  | "draft"
  | "submitted"
  | "need_supplement"
  | "finance_review"
  | "approved"
  | "rejected";

export type AttachmentType = "invoice" | "payment" | "approval" | "other";

export type AuditResult = "pass" | "attention" | "risk";

export interface ExpenseAttachment {
  id: string;
  claim_id: string;
  file_uri: string;
  file_name: string;
  file_type: string;
  size?: number | null;
  attachment_type: AttachmentType;
  ocr_result?: Record<string, unknown> | null;
  extracted_fields?: Record<string, unknown> | null;
  ocr_confidence?: string | number | null;
  classification_source?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ExpenseAuditItem {
  id: string;
  claim_id: string;
  name: string;
  result: AuditResult;
  evidence?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ExpenseClaim {
  id: string;
  user_id: string;
  status: ExpenseStatus;
  claim_no?: string | null;
  expense_type: string;
  title?: string | null;
  description?: string | null;
  total_amount?: string | number | null;
  currency: string;
  submitted_at?: string | null;
  approved_at?: string | null;
  audit_summary?: {
    level?: AuditResult;
    summary?: string;
    missing_materials?: string[];
    next_action?: string;
    metadata?: Record<string, unknown>;
  } | null;
  reviewer_id?: string | null;
  reviewed_at?: string | null;
  review_comment?: string | null;
  created_at: string;
  updated_at: string;
  attachments: ExpenseAttachment[];
  audit_items: ExpenseAuditItem[];
}

export interface ExpenseApprovalLog {
  id: string;
  claim_id: string;
  actor_id: string;
  action: "submit" | "resubmit" | "approve" | "reject" | "request_supplement";
  from_status?: string | null;
  to_status?: string | null;
  comment?: string | null;
  snapshot?: Record<string, unknown> | null;
  created_at: string;
}

export interface TravelExpenseStandard {
  id: string;
  name: string;
  org_id?: string | null;
  city_tier?: string | null;
  daily_limit?: string | number | null;
  single_trip_limit?: string | number | null;
  currency: string;
  effective_from?: string | null;
  effective_to?: string | null;
  is_active: boolean;
  metadata_?: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
}
